resources/aws/lib/opentelemetry/resource/detector/aws/ec2.rb

```diff
@@ -0,0 +1,137 @@
+              resource_attributes[OpenTelemetry::SemanticConventions::Resource::HOST_NAME] = hostname
+            rescue StandardError => e
+              OpenTelemetry.logger.debug("EC2 resource detection failed: #{e.message}")
+              return OpenTelemetry::SDK::Resources::Resource.create({})
+            end
+
+            # Filter out nil or empty values
+            resource_attributes.delete_if { |_key, value| value.nil? || value.empty? }
+            OpenTelemetry::SDK::Resources::Resource.create(resource_attributes)
+          end
+
+          private
+          
+          # Fetches an IMDSv2 token from the EC2 metadata service
+          #
+          # @return [String, nil] The token or nil if the request failed
+          def fetch_token
+            uri = URI.parse("http://#{EC2_METADATA_HOST}#{TOKEN_ENDPOINT}")
+            request = Net::HTTP::Put.new(uri)
+            request[TOKEN_TTL_HEADER] = TOKEN_TTL_VALUE
+
+            response = make_request(uri, request)
+            return nil unless response.is_a?(Net::HTTPSuccess)
+
+            response.body
+          end
+
+          # Fetches the instance identity document which contains EC2 instance metadata
+          #
+          # @param token [String] IMDSv2 token
+          # @return [Hash, nil] Parsed identity document or nil if the request failed
+          def fetch_identity_document(token)
+            uri = URI.parse("http://#{EC2_METADATA_HOST}#{IDENTITY_DOCUMENT_ENDPOINT}")
+            request = Net::HTTP::Get.new(uri)
+            request[TOKEN_HEADER] = token
+
+            response = make_request(uri, request)
+            return nil unless response.is_a?(Net::HTTPSuccess)
+
+            begin
+              JSON.parse(response.body)
+            rescue JSON::ParserError
+              nil
+            end
+          end
+
+          # Fetches the EC2 instance hostname
+          #
+          # @param token [String] IMDSv2 token
+          # @return [String, nil] The hostname or nil if the request failed
+          def fetch_hostname(token)
+            uri = URI.parse("http://#{EC2_METADATA_HOST}#{HOSTNAME_ENDPOINT}")
+            request = Net::HTTP::Get.new(uri)
+            request[TOKEN_HEADER] = token
+
+            response = make_request(uri, request)
+            return nil unless response.is_a?(Net::HTTPSuccess)
+
+            response.body
+          end
+
+          # Makes an HTTP request with timeout handling
+          #
+          # @param uri [URI] The request URI
+          # @param request [Net::HTTP::Request] The request to perform
+          # @return [Net::HTTPResponse, nil] The response or nil if the request failed
+          def make_request(uri, request)
+            http = Net::HTTP.new(uri.host, uri.port)
+            http.open_timeout = HTTP_TIMEOUT
+            http.read_timeout = HTTP_TIMEOUT
+
+            begin
+              http.request(request)
```

Comment:
Add `OpenTelemetry::Common::Utilities.untraced` around http.request to avoid random span from detection.